Updated version skew strategy for InPlacePodVerticalScaling

sreeram-venkitesh opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature

#### What this PR does / why we need it:

Updates the version skew strategy for InPlacePodVerticalScaling for beta graduation.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #117767

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Version skew strategy update for InPlacePodVerticalScaling for beta graduation.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/1287
```

sreeram-venkitesh commented:
@tallclair Please let me know if I’m in the right direction. Some tests are breaking, I’m still working on fixing them. This is how I tested the change:

- Create kind cluster with 3 nodes: 
  - control plane with `kindest/node:latest` image and InPlacePodVerticalScaling enabled
  - worker # 1 with `kindest/node:v1.31.0` with InPlacePodVerticalScaling disabled
  - worker # 2 with `kindest/node:v1.30.0` with  InPlacePodVerticalScaling disabled

Previous to these changes, I was able to update Pod's resources. In v1.30, the Pod gets restarted, and in v1.31, the Pod gets updated, but its not restarted so the resource update is not reflected until the next restart. After the changes I made in this PR, Pod resource update is not allowed for Pods created on both the worker nodes with InPlacePodVerticalScaling disabled. The validation checks if the Pod status has resources defined before update and I’ve used the same logic you wrote in the KEP.  

This is how the error looks like right now when you try to update a Pod resource without enabling the feature gate. I’m also getting the error for how Resources cannot be updated along with this. I’m assuming this is because `cmp.Diff(oldPod.Spec, mungedPodSpec)` is non-empty. I’m looking into this as well. Also the string `Pod running on node without IPPVS enabled may not be updated` is a placeholder, I will change this.

```yaml
# pods "nginx-pod" was not valid:
# * spec: Forbidden: Pod running on node without IPPVS enabled may not be updated
# * spec: Forbidden: pod updates may not change fields other than `spec.containers[*].image`,`spec.initContainers[*].image`,`spec.activeDeadlineSeconds`,`spec.tolerations` (only additions to existing tolerations),`spec.terminationGracePeriodSeconds` (allow it to be set to 1 if it was previously negative)
#   core.PodSpec{
#       Volumes:        {{Name: "kube-api-access-njhml", VolumeSource: {Projected: &{Sources: {{ServiceAccountToken: &{ExpirationSeconds: 3607, Path: "token"}}, {ConfigMap: &{LocalObjectReference: {Name: "kube-root-ca.crt"}, Items: {{Key: "ca.crt", Path: "ca.crt"}}}}, {DownwardAPI: &{Items: {{Path: "namespace", FieldRef: &{APIVersion: "v1", FieldPath: "metadata.namespace"}}}}}}, DefaultMode: &420}}}},
#       InitContainers: nil,
#       Containers: []core.Container{
#               {
#                       ... // 6 identical fields
#                       EnvFrom: nil,
#                       Env:     nil,
#                       Resources: core.ResourceRequirements{
#                               Limits: nil,
#                               Requests: core.ResourceList{
#                                       s"cpu":    {i: {...}, s: "10m", Format: "DecimalSI"},
# -                                     s"memory": {i: resource.int64Amount{value: 104857600}, s: "100Mi", Format: "BinarySI"},
# +                                     s"memory": {i: resource.int64Amount{value: 105906176}, s: "101Mi", Format: "BinarySI"},
#                               },
#                               Claims: nil,
#                       },
#                       ResizePolicy:  {{ResourceName: s"cpu", RestartPolicy: "NotRequired"}, {ResourceName: s"memory", RestartPolicy: "NotRequired"}},
#                       RestartPolicy: nil,
#                       ... // 13 identical fields
#               },
#       },
#       EphemeralContainers: nil,
#       RestartPolicy:       "Always",
#       ... // 28 identical fields
#   }
```

tallclair commented:
/assign

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5170,6 +5170,11 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
+	isPodResizeRequestValid := isPodResizeRequestValid(*oldPod)
+	if !isPodResizeRequestValid {
+		allErrs = append(allErrs, field.Forbidden(specPath, "Pod running on node without IPPVS enabled may not be updated"))
+	}
```
This will probably be OK once we switch to the resize subresource (https://github.com/kubernetes/kubernetes/issues/109553), but to call it here you'll need to confirm it's actually a resize request (there are other valid pod updates, including metadata & image)

sreeram-venkitesh replied:
@tallclair In order to test this, I created a new branch from Anish's PR and added these changes to the new `ValidatePodResize` function that is added. This way this validation error would be added only to resize requests and all the other tests wouldn't fail like in this PR. Once Anish's PR gets merged I can move these changes to the `ValidatePodResize` function too. I opened https://github.com/kubernetes/kubernetes/pull/128540 to test this first and be ready before code freeze. I'll close it once I'm done testing.

tallclair replied:
Yes, agreed - once Anish's PR merges this should just move to `ValidatePodResize`, and you don't need to worry about the feature gate.

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5178,7 +5183,7 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 		// When the feature-gate is turned off, any new requests attempting to update CPU or memory
 		// resource values will result in validation failure.
-		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) && isPodResizeRequestValid {
```
I think skip this check here. The error will already be added above, so flagging resource mutation as invalid here will just make for a confusing and redundant error message.

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5170,6 +5170,11 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 		allErrs = append(allErrs, field.Invalid(fldPath, newPod.Status.QOSClass, "Pod QoS is immutable"))
 	}
 
+	isPodResizeRequestValid := isPodResizeRequestValid(*oldPod)
```
this might need to move to `PodValidationOptions`. TBH, I'm not sure what decides what goes in the validation options - maybe ask an API reviewer.

sreeram-venkitesh replied:
I was trying to move this `isPodResizeRequestValid` function to PodValidationOptions and make it aligned with everything else there, but I soon realized that the opts that `ValidatePodUpdate` gets comes from the `GetValidationOptionsFromPodSpecAndMeta` function. We need to check if Resources are present in the Status however. Can PodValidationOptions have logic based on the PodStatus?

https://github.com/kubernetes/kubernetes/blob/8c7160205db67b250dece7cad5c6367debad2642/pkg/registry/core/pod/strategy.go#L149-L151

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5290,6 +5295,19 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
+	if !utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		return false
+	}
+	for _, c := range pod.Status.ContainerStatuses {
```
Add a comment here explaining what this is doing and why (i.e. for version skew, and why it works)

tallclair replied:
Also, leave a TODO to remove this code after GA+3 releases.

tallclair reviewed (commented):
On the right track. Thanks!

tallclair reviewed (commented):
I think this is good to go once you rebase it on the /resize subresource PR.

Have you manually tested the changes? If not, I can share some instructions for setting up a version-skewed Kind cluster to test locally.

sreeram-venkitesh commented:
@tallclair I've rebased my changes and squashed everything into a single commit, PTAL!

> Have you manually tested the changes?

I had tested the skew before the rebase as described in https://github.com/kubernetes/kubernetes/pull/128186#issuecomment-2423180494. I will test it once again today.

sreeram-venkitesh committed:
Updated version skew strategy for InPlacePodVerticalScaling

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5546,6 +5552,25 @@ func ValidatePodResize(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
+	// if the Pod already has Pod.Status.ContainerStatuses[i].Resources. This means
+	// that the apiserver would only allow updates to Pods running on Nodes with
+	// the InPlacePodVerticalScaling feature gate enabled.
+	if !utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		return false
+	}
```
ValidatePodResize is never called if the feature gate isn't enabled, so no need to check it here.

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5546,6 +5552,25 @@ func ValidatePodResize(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 	return allErrs
 }
 
+func isPodResizeRequestValid(pod core.Pod) bool {
```
```suggestion
// isPodResizeRequestSupported checks whether the pod being is running on a node with InPlacePodVerticalScaling enabled. 
func isPodResizeRequestSupported(pod core.Pod) bool {
```

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5546,6 +5552,25 @@ func ValidatePodResize(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
+			return c.Resources != nil
+		}
+	}
+	// No running containers
```
```suggestion
	// No running containers. We cannot tell whether the node supports resize at this point, so we assume it does.
```

tallclair reviewed (commented):
Thanks!

sreeram-venkitesh committed:
Update function name and remove feature gate check

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5506,6 +5506,12 @@ func ValidatePodResize(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 		allErrs = append(allErrs, field.Invalid(specPath, newPod.Status.QOSClass, "Pod QOS Class may not change as a result of resizing"))
 	}
 
+	isPodResizeRequestValid := isPodResizeRequestSupported(*oldPod)
+
+	if !isPodResizeRequestValid {
```
nit: no need for the variable
```suggestion
	if !isPodResizeRequestSupported(*oldPod) {
```

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5506,6 +5506,12 @@ func ValidatePodResize(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
+	isPodResizeRequestValid := isPodResizeRequestSupported(*oldPod)
+
+	if !isPodResizeRequestValid {
+		allErrs = append(allErrs, field.Forbidden(specPath, "Pod running on node without InPlacePodVerticalScaling feature gate enabled may not be updated"))
```
```suggestion
		allErrs = append(allErrs, field.Forbidden(specPath, "Pod running on node without support for resize"))
```

tallclair commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -25094,95 +25094,126 @@ func TestValidatePodResize(t *testing.T) {
+		test                            string
+		old                             *core.Pod
+		new                             *core.Pod
+		enableInPlacePodVerticalScaling bool
```
nit: remove this

tallclair commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -25094,95 +25094,126 @@ func TestValidatePodResize(t *testing.T) {
 	}
 
 	for _, test := range tests {
+		featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.InPlacePodVerticalScaling, test.enableInPlacePodVerticalScaling)
+
```
no longer necessary

tallclair reviewed (commented):
Looks like the unit test needs to be fixed too.

tallclair commented:
/triage accepted
/priority important-soon
/milestone v1.32

sreeram-venkitesh committed:
Fixes from review, updated tests cases

sreeram-venkitesh committed:
Removed duplicated tests after rebase

sreeram-venkitesh committed:
Fix verify-gofmt.sh

tallclair commented:
Wait, where'd the unit tests go? We don't need to change the feature gate, since this function isn't checking the feature gate, but we still want to manually test the logic. I'd expect to see test cases with & without running containers, and with & without status resources for each.

sreeram-venkitesh commented:
I'll push them now.

sreeram-venkitesh committed:
Added unit tests

tallclair commented:
/test pull-e2e-gci-gce-alpha-enabled-default

tallclair commented:
/lgtm

/assign @thockin 
For validation approval.

tallclair commented:
Test failures look like unrelated flakes. That suite isn't pr blocking.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5515,6 +5515,10 @@ func ValidatePodResize(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 		allErrs = append(allErrs, field.Invalid(specPath, newPod.Status.QOSClass, "Pod QOS Class may not change as a result of resizing"))
 	}
 
+	if !isPodResizeRequestSupported(*oldPod) {
```
@ndixita you have similar logic (different reason, same result) - PTAL and see if we can join them, whoever merges last :)

thockin reviewed (approved):
Thanks!

/lgtm
/approve

pacoxu commented:
/milestone v1.32
approved before code freeze
/skip
